src/aggregate_with_randomness.cc

```diff
@@ -0,0 +1,247 @@
+    ret_val.Set(
+        Napi::String::New(env, "publicKey"),
+        module->public_key_ctr.New({Napi::External<blst_ts::P1Wrapper>::New(
+            env, new blst_ts::P1{std::move(aggregate_key)})}));
+    ret_val.Set(
+        Napi::String::New(env, "signature"),
+        module->signature_ctr.New({Napi::External<blst_ts::P2Wrapper>::New(
+            env, new blst_ts::P2{std::move(aggregate_sig)})}));
+    return scope.Escape(ret_val);
+}
+
+class AggregateWithRandomnessWorker : public Napi::AsyncWorker {
+   public:
+    AggregateWithRandomnessWorker(
+        const Napi::CallbackInfo &info,
+        BlstTsAddon *module,
+        std::vector<SignatureAndPublicKeySet> sets,
+        size_t sets_length,
+        bool validate)
+        : Napi::AsyncWorker{info.Env(), "AggregateWithRandomnessWorker"},
+          deferred{Env()},
+          _module{std::move(module)},
+          _sets_ref{Napi::Persistent(info[0])},
+          _sets{std::move(sets)},
+          _sets_length{std::move(sets_length)},
+          _validate{std::move(validate)} {}
+
+    /**
+     * GetPromise associated with deferred for return to JS
+     */
+    Napi::Promise GetPromise() { return deferred.Promise(); }
+
+   protected:
+    void Execute() {
+        std::string error_msg{};
+        blst_ts::BLST_TS_ERROR err = aggregate_with_randomness(
+            _aggregate_key,
+            _aggregate_sig,
+            error_msg,
+            _module,
+            _sets,
+            _sets_length,
+            _validate);
+        if (err == blst_ts::BLST_TS_ERROR::HAS_NATIVE_ERROR) {
+            SetError(error_msg);
+        }
+    }
+    void OnOK() {
+        Napi::Env env = Env();
+        Napi::Object ret_val = Napi::Object::New(env);
+        ret_val["publicKey"] = _module->public_key_ctr.New(
+            {Napi::External<blst_ts::P1Wrapper>::New(
+                env, new blst_ts::P1{std::move(_aggregate_key)})});
+        ret_val["signature"] =
+            _module->signature_ctr.New({Napi::External<blst_ts::P2Wrapper>::New(
+                env, new blst_ts::P2{std::move(_aggregate_sig)})});
+        deferred.Resolve(ret_val);
+    }
+    void OnError(const Napi::Error &err) { deferred.Reject(err.Value()); }
+
+   public:
+    Napi::Promise::Deferred deferred;
+
+   private:
+    BlstTsAddon *_module;
+    Napi::Reference<Napi::Value> _sets_ref;
+    std::vector<SignatureAndPublicKeySet> _sets;
+    size_t _sets_length;
+    bool _validate;
+    blst::P1 _aggregate_key;
+    blst::P2 _aggregate_sig;
+};
+
+Napi::Value AsyncAggregateWithRandomness(const Napi::CallbackInfo &info) {
+    BLST_TS_FUNCTION_PREAMBLE(info, env, module)
+    std::vector<SignatureAndPublicKeySet> sets{};
+    size_t sets_length{0};
+    bool validate{true};
+
+    blst_ts::BLST_TS_ERROR err =
+        prepare_aggregate_with_randomness(sets, sets_length, validate, info);
+    if (err == blst_ts::BLST_TS_ERROR::JS_ERROR_THROWN) {
+        return env.Undefined();
+    }
```

Comment:
If it is another kind of error, is it fine to continue? Or should we continue only on no error at all?

Reply:
No other status codes can be generated from that function other than `OK` and `JS_ERROR_THROWN`.  The native errors that can potentially come up are wrapped in try/catch and a JS error is thrown